### ISSUE TITLE
Consider an O(ND) Difference Algorithm for DOM tree merging

### ISSUE THREAD

**Issue by jonathantneal** [atomicojs/atomico#8]
I tried your diffing algorithm and experienced the recreation/reloading of images and iframe elements whenever others elements were prepended before them. This can happen frequently when creating filter-able lists that contain media.

To resolve this, check out the [O(ND) Difference Algorithm and Its Variations](http://citeseerx.ist.psu.edu/viewdoc/download?doi=10.1.1.4.6927&rep=rep1&type=pdf) as well as these [pre-optimizations](https://neil.fraser.name/writing/diff/). Alternatively, consider leveraging a third-party library like [`domdiff`](https://github.com/WebReflection/domdiff).

I can write a patcher for you over domdiff, but it would come in around 1.52 kB min/gzip — which isn’t too bad when [domdiff is 1.5 kB min/gzip by itself](https://bundlephobia.com/result?p=domdiff@2.0.3).

**Comment by UpperCod** [comment]
Hi, Thanks for your issue, I have some questions.

1. Was the process hydration ?, Since this is not completely complete, currently only concerns the node, but some attributes redefine it
2. Do I use keyes?
3. You may attach the example, I have generated a similar one in [codesanbox](https://codesandbox.io/s/k54n17y5lv) I hope it reflects your problem

by the way the work of @WebReflection is incredible, I have analyzed its code as part of my learning

I have analyzed it and I have not used domdiff for this.

1. HoCs : adding support for this pattern is expensive in lines of code within domdiff.
2. management of attributes and properties: by default this does not include
3. children: atomico treats children as React, this is to apply the default memo pattern.
4. shadowDom: atomic allows you to apply shadowDom as an attribute, this in domdiff would break the scanning of children
5. tag `<host />`, this tag is designed to facilitate work with WC.
```jsx
class Tag extends Element {
	/**❌**/
	constructor() {
		super();
		this.attachShadow({ mode: "open" });
		this.style.background = "black";
		this.addEventListener("click", () => {
			console.log("event!");
		});
	}
	/**✔️**/
	render() {
		return (
			<host
				shadowDom
				onClick={() => console.log("event")}
				style={{ background: "black" }}
			/>
		);
	}
}
```

**Comment by UpperCod** [comment]
Hi!, Identify an error in children's patching with key, adding more children in random order... It has been fixed in version 1.4.1 [commit](https://github.com/atomicojs/core/commit/0601ed2ae8f09d8a8bafdd5e9769986293457166).  [test](https://github.com/atomicojs/core/blob/master/test/children/keyes.test.js#L51-L124)
